kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright (2023) The Delta Lake Project Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.delta.kernel.internal;
+
+import java.io.IOException;
+import java.nio.file.FileAlreadyExistsException;
+import java.util.*;
+
+import io.delta.kernel.*;
+import io.delta.kernel.client.TableClient;
+import io.delta.kernel.data.Row;
+import io.delta.kernel.types.StructType;
+import io.delta.kernel.utils.CloseableIterable;
+import io.delta.kernel.utils.CloseableIterator;
+
+import io.delta.kernel.internal.actions.*;
+import io.delta.kernel.internal.data.TransactionStateRow;
+import io.delta.kernel.internal.fs.Path;
+import io.delta.kernel.internal.util.FileNames;
+import io.delta.kernel.internal.util.VectorUtils;
+import static io.delta.kernel.internal.TableConfig.CHECKPOINT_INTERVAL;
+import static io.delta.kernel.internal.actions.SingleAction.*;
+import static io.delta.kernel.internal.util.Preconditions.checkState;
+import static io.delta.kernel.internal.util.Utils.toCloseableIterator;
+
+public class TransactionImpl implements Transaction {
+    public static final int DEFAULT_READ_VERSION = 1;
+    public static final int DEFAULT_WRITE_VERSION = 2;
+
+    private final UUID txnId = UUID.randomUUID();
+
+    private final String engineInfo;
+    private final Operation operation;
+    private final Path dataPath;
+    private final Path logPath;
+    private final Optional<StructType> newSchema;
+    private final Optional<List<String>> newPartitionColumns;
```

Comment:
Does new imply that the partition columns can be changed?

Reply:
Maybe just `partitionColumnsOpt`?

Reply:
I refactored the code. We don't this anymore. The snapshot should contain schema or partition columns either from the InitialSnapshot or Snapshot (loaded from the table)